ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java

```diff
@@ -516,20 +514,11 @@ public LogEntryProto getEntry(TimeDuration timeout) throws RaftLogIOException, T
       if (LogProtoUtils.isStateMachineDataEmpty(entryProto)) {
         final String err = getName() + ": State machine data not set for " + this;
         LOG.error(err);
-        data.release();
         throw new RaftLogIOException(err);
```

Comment:
For this exception, I was thinking of such a scenario:

1. In `List<LogEntryProto> protos = buffer.pollList(...)`, if `buffer` has 10 records, and `pollList` has finished polling 7 records, on the 8th record, it reaches this RaftLogIOException().
2. In outer finally, the leftover 3 records will be cleaned and removed, but the polled 7 records won't be released.

Reply:
@duongkame , we should also catch `RaftLogIOException` and release all the entries in that case.
```java
    final List<LogEntryProto> protos;
    try {
      protos = buffer.pollList(getHeartbeatWaitTimeMs(), EntryWithData::getEntry,
          (entry, time, exception) -> LOG.warn("Failed to get {} in {}",
              entry, time.toString(TimeUnit.MILLISECONDS, 3), exception));
    } catch (RaftLogIOException e) {
      for (ReferenceCountedObject<EntryWithData> ref : offered.values()) {
        ref.release();
      }
      offered.clear();
      throw e;
    } finally {
      for (EntryWithData entry : buffer) {
        // Release remaining entries.
        offered.remove(entry.getIndex()).release();
      }
      buffer.clear();
    }
```

Reply:
Thanks @szetszwo  @symious .